metadata/src/main/java/org/apache/kafka/controller/QuorumController.java

```diff
@@ -0,0 +1,920 @@
+        Map<ConfigResource, Map<String, String>> newConfigs, boolean validateOnly) {
+        return appendWriteEvent("legacyAlterConfigs", () -> {
+            ControllerResult<Map<ConfigResource, ApiError>> result =
+                configurationControl.legacyAlterConfigs(newConfigs);
+            if (validateOnly) {
+                return result.withoutRecords();
+            } else {
+                return result;
+            }
+        });
+    }
+
+    class ProcessBrokerHeartbeatOperation
+            implements ControllerWriteOperation<BrokerHeartbeatReply> {
+        private final BrokerHeartbeatRequestData request;
+        private boolean updateShutdownOffset = false;
+
+        ProcessBrokerHeartbeatOperation(BrokerHeartbeatRequestData request) {
+            this.request = request;
+        }
+
+        @Override
+        public ControllerResult<BrokerHeartbeatReply> generateRecordsAndResult() {
+            clusterControl.checkBrokerEpoch(request.brokerId(), request.brokerEpoch());
+            List<ApiMessageAndVersion> records = new ArrayList<>();
+            boolean movingLeadersForShutDown = false;
+            if (request.wantShutDown()) {
+                records = replicationControl.removeLeaderships(request.brokerId());
+                movingLeadersForShutDown = !records.isEmpty();
+            }
+            ControllerResult<BrokerHeartbeatReply> result = clusterControl.
+                processBrokerHeartbeat(request, lastCommittedOffset,
+                    movingLeadersForShutDown);
+            if (movingLeadersForShutDown && !result.response().shouldShutDown()) {
+                updateShutdownOffset = true;
+            }
+            if (result.response().isFenced()) {
+                replicationControl.removeFromIsr(request.brokerId(), records);
+            }
+            rescheduleMaybeFenceReplicas();
+            records.addAll(result.records());
+            return new ControllerResult<>(records, result.response());
+        }
+
+        @Override
+        public void processBatchEndOffset(long offset) {
+            if (updateShutdownOffset) {
+                clusterControl.updateShutdownOffset(request.brokerId(), offset);
+            }
+        }
+    }
+
+    @Override
+    public CompletableFuture<BrokerHeartbeatReply>
+            processBrokerHeartbeat(BrokerHeartbeatRequestData request) {
+        return appendWriteEvent("processBrokerHeartbeat",
+            new ProcessBrokerHeartbeatOperation(request));
+    }
+
+    @Override
+    public CompletableFuture<BrokerRegistrationReply>
+            registerBroker(BrokerRegistrationRequestData request) {
+        return appendWriteEvent("registerBroker", () -> {
+            ControllerResult<BrokerRegistrationReply> result = clusterControl.
+                registerBroker(request, writeOffset + 1, featureControl.
+                    finalizedFeaturesAndEpoch(Long.MAX_VALUE));
```

Comment:
I believe after this we need to invoke something to cover the case where a topic has this broker as its only ISR member:

```
replicationControl.maybeMakeLeader(request.brokerId(), result.records());
```

The implementation might look like this:

```
    void maybeMakeLeader(int brokerId, List<ApiMessageAndVersion> records) {
        Iterator<TopicPartition> iterator = brokersToIsrs.iterator(brokerId, false);
        while (iterator.hasNext()) {
            TopicPartition topicPartition = iterator.next();
            TopicControlInfo topic = topics.get(topicPartition.topicId());
            if (topic == null) {
                throw new RuntimeException("Topic ID " + topicPartition.topicId() + " existed in " +
                    "isrMembers, but not in the topics map.");
            }
            PartitionControlInfo partition = topic.parts.get(topicPartition.partitionId());
            if (partition == null) {
                throw new RuntimeException("Partition " + topicPartition +
                    " existed in isrMembers, but not in the partitions map.");
            }
            // make it the leader if it is the only ISR member
            if (partition.isr.length == 1 && partition.isr[0] == brokerId && partition.leader != brokerId) {
                int newLeader = brokerId;
                int newLeaderEpoch = partition.leaderEpoch + 1;
                IsrChangeRecord message = new IsrChangeRecord().
                    setPartitionId(topicPartition.partitionId()).
                    setTopicId(topic.id).
                    setIsr(Replicas.toList(partition.isr)).
                    setLeader(newLeader).
                    setLeaderEpoch(newLeaderEpoch).
                    setPartitionEpoch(partition.partitionEpoch + 1);
                records.add(new ApiMessageAndVersion(message, (short) 0));
            }
        }
    }
```

Reply:
Thanks, @rondagostino .  In the latest version, I do not remove brokers from singleton ISRs.